### ISSUE TITLE
Set up prodigy with the appropriate NER and llm recipe

### ISSUE THREAD

**Issue by india-kerle** [nestauk/dap_prinz_green_jobs#44]
I think we actually want to set up prodigy with a different recipe to make use of llms! I've added slides  that describes these steps.

**Comment by india-kerle** [comment]
I think we actually want to set it up like:

prodigy ner.openai.fetch examples.jsonl openai-out.jsonl skill,company_description,qualification -F ./recipes/ner.py

**Comment by india-kerle** [comment]
as described in slide 9 here: https://docs.google.com/presentation/d/1qChjGIxmDYYBJb4gWEi1oZUd-xj2rtrclFKi9T0McSI/edit#slide=id.g250cf64d1cb_0_0 

although they use example entities ^^

**Comment by india-kerle** [comment]
We made a decision on the entities we want to extract in this issue (#33) + we would likely want to collect more skill/multiskills as well for the skills side of things: 

1. company description
2. company sector
3. qualification
4. skill 
5. multiskill
6. company benefit

**Comment by india-kerle** [comment]
From a labelling data perspective, I reckon it would be best to label the mixed sample to make sure there's representation of green jobs - s3://prinz-green-jobs/outputs/data/ojo_application/deduplicated_sample/mixed_ojo_sample.csv 

UPDATE: I've converted this dataset to a .jsonl file, as recommended by Prodigy here: s3://prinz-green-jobs/inputs/data/training_data/mixed_ojo_sample.jsonl

**Comment by india-kerle** [comment]
Ok i've started looking into this. I've:

1) installed prodigy 
2) `git clone git@github.com:explosion/prodigy-openai-recipes.git`
3) saved training data .jsonl files from s3 (all files in `s3://prinz-green-jobs/inputs/data/training_data/`) to the data directory in the cloned repo
4) added openAI org name and API key in .env file in the root directory (using @beingkk 's for the time being for testing)
5) test set up with example python command in repo

```
python -m prodigy ner.openai.correct my_ner_data ./data/reddit_r_cooking_sample.jsonl "dish,ingredient,equipment" -p ./templates/ner_prompt.jinja2 -e ./examples/ner.yaml -n 2 -F ./recipes/openai_ner.py
```

(it ran for me with these instructions)

6) run the following (at the root of the cloned repo): 

```
python -m prodigy ner.openai.correct mixed_ojo_sample_mini ./data/mixed_ojo_sample_mini.jsonl "COMPANY_DESCRIPTION,COMPANY_SECTOR,QUALIFICATION,SKILL,MULTISKILL,COMPANY_BENEFIT" -p ./templates/ner_prompt.jinja2  -F ./recipes/openai_ner.py
```

This results in a `httpx.ReadTimeout: The read operation timed out` error. I think it could have to do with the size of the job adverts? we could: make a smaller training dataset (mini=500 job adverts, small=1000, normal=10000)

**Comment by lizgzil** [comment]
I am getting
```
Retrying call (retries left: 10, timeout: 20s). Previous call returned: 429 (Too
Many Requests)
Retrying call (retries left: 9, timeout: 40s). Previous call returned: 429 (Too
Many Requests)
Retrying call (retries left: 8, timeout: 80s). Previous call returned: 429 (Too
Many Requests)
```
even when using the example dataset! I am using my own openai keys though

**Comment by india-kerle** [comment]
Ok I was able to get it to work with just 1 advert (to avoid the read time out error, which I still received with a data sample of 5 job postings) and saved the annotations as an example. 

![Screenshot 2023-06-22 at 17 06 03](https://github.com/nestauk/dap_prinz_green_jobs/assets/46863334/981e7675-30a4-493d-966b-e5a02a621e41)

A few initial thoughts:

1. timeouts - I _think_ its because the job advert text is very long - having a google, looks like there have been a lot of API timeout errors using the gpt-3.5-turbo model and text-davinci-003 (prodigy's default) - we might want to slightly customise the recipe with better API timeout handling (I think there is a ton on this online) + pass a longer timeout time in openai_ner.py? 

2. prompts - from this initial example, it looks like it will get as many i.e. company descriptions as possible in the text - we may want to tune the prompt to only include 1 company description, or at least cap it? Labels most skills as multiskills and single worded skills as skill  

3. Training data - I think we will want to clean up the job advert text _before_ labelling...

**Comment by lizgzil** [comment]
great you got it working! the ingredients example finally worked for me too. 

Not sure about the timeouts, but something to help this might be to add a `./examples/ner.yaml` file for this task so it's not zero-shot learning?

**Comment by india-kerle** [comment]
oooh nice! I'll give that a go!

**Comment by india-kerle** [comment]
okay live updates: the kicker is using Nesta's wifi for this task - the timeout doesn't happen when I'm on my home wifi - I've had issues with calling APIs in the past on Nesta's wifi

**Comment by india-kerle** [comment]
closing this given we've determined a different approach to prodigy recipes
